File: domain/relationship.ts
/**
 * domain/relationship.ts
 *
 * Relationship is the pure domain model for a typed, evidence-backed edge in
 * the knowledge graph connecting two Entity references. A Relationship
 * carries its own confidence and optional temporal validity window,
 * independent of the confidence of either endpoint entity.
 *
 * Mirrors contracts/domain/relationship.schema.json field-for-field.
 *
 * Domain Purity: standard runtime and sibling domain modules only.
 */

import {
  Status,
  freezeRefs,
  freezeTags,
  newId,
  now,
  type Id,
  type Metadata,
  type Provenance,
  type Reference,
  type Timestamp,
  EMPTY_METADATA,
} from "./common.js";
import { ClassificationLevel } from "./classification_level.js";
import { ConfidenceScore } from "./confidence_score.js";
import { DomainInvariantError } from "./common.js";

export enum RelationshipType {
  AssociatedWith = "associated_with",
  Owns = "owns",
  EmployedBy = "employed_by",
  LocatedAt = "located_at",
  CommunicatesWith = "communicates_with",
  Controls = "controls",
  MemberOf = "member_of",
  TransactedWith = "transacted_with",
  Other = "other",
}

export interface RelationshipInit {
  readonly relationshipType: RelationshipType;
  readonly sourceEntityRef: Reference;
  readonly targetEntityRef: Reference;
  readonly directional?: boolean;
  readonly validFrom?: Timestamp;
  readonly validTo?: Timestamp;
  readonly status?: Status;
  readonly classification?: ClassificationLevel;
  readonly evidenceRefs?: ReadonlyArray<Reference>;
  readonly sourceRefs?: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags?: ReadonlyArray<string>;
  readonly metadata?: Metadata;
}

/** Relationship is an immutable, typed edge between two entity references. */
export class Relationship {
  readonly id: Id;
  readonly version: string;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
  readonly status: Status;
  readonly classification: ClassificationLevel;
  readonly relationshipType: RelationshipType;
  readonly sourceEntityRef: Reference;
  readonly targetEntityRef: Reference;
  readonly directional: boolean;
  readonly validFrom?: Timestamp;
  readonly validTo?: Timestamp;
  readonly evidenceRefs: ReadonlyArray<Reference>;
  readonly sourceRefs: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags: ReadonlyArray<string>;
  readonly metadata: Metadata;

  private constructor(id: Id, version: string, createdAt: Timestamp, updatedAt: Timestamp, init: RelationshipInit) {
    if (init.sourceEntityRef.refId === init.targetEntityRef.refId) {
      throw new DomainInvariantError("Relationship.sourceEntityRef and targetEntityRef must reference distinct entities.");
    }
    if (init.validFrom !== undefined && init.validTo !== undefined && init.validTo < init.validFrom) {
      throw new DomainInvariantError("Relationship.validTo must not precede validFrom.");
    }
    this.id = id;
    this.version = version;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.status = init.status ?? Status.Active;
    this.classification = init.classification ?? ClassificationLevel.Internal;
    this.relationshipType = init.relationshipType;
    this.sourceEntityRef = init.sourceEntityRef;
    this.targetEntityRef = init.targetEntityRef;
    this.directional = init.directional ?? true;
    this.validFrom = init.validFrom;
    this.validTo = init.validTo;
    this.evidenceRefs = freezeRefs(init.evidenceRefs);
    this.sourceRefs = freezeRefs(init.sourceRefs);
    this.confidence = init.confidence;
    this.provenance = init.provenance;
    this.tags = freezeTags(init.tags);
    this.metadata = init.metadata ?? EMPTY_METADATA;
    Object.freeze(this);
  }

  static create(init: RelationshipInit): Relationship {
    const timestamp = now();
    return new Relationship(newId(), "1.0.0", timestamp, timestamp, init);
  }

  /** True if `entityRef` is either endpoint of this relationship. */
  involves(entityRef: Reference): boolean {
    return this.sourceEntityRef.refId === entityRef.refId || this.targetEntityRef.refId === entityRef.refId;
  }

  /** True if this relationship's validity window includes the given timestamp. */
  isValidAt(timestamp: Timestamp): boolean {
    if (this.validFrom !== undefined && timestamp < this.validFrom) return false;
    if (this.validTo !== undefined && timestamp > this.validTo) return false;
    return true;
  }
}

File: domain/relationship.py
"""
domain/relationship.py

Relationship domain model - connections between entities.
Mirrors contracts/domain/relationship.schema.json field-for-field.
"""
from dataclasses import dataclass, field
from enum import Enum

from .classification_level import ClassificationLevel, ClassificationLevelValue
from .confidence_score import ConfidenceScore, ConfidenceLevel, ConfidenceMethods
from .common import (



    EMPTY_METADATA,
    Metadata,
    Provenance,
    Reference,
    Status,
    new_id,
    utc_now,
)


class RelationshipType(str, Enum):
    """Types of relationships between entities."""
    OWNS = "owns"
    CONTROLS = "controls"
    ASSOCIATED_WITH = "associated_with"
    LOCATED_AT = "located_at"
    WORKS_FOR = "works_for"
    COMMUNICATES_WITH = "communicates_with"
    TRANSACTS_WITH = "transacts_with"
    HOSTS = "hosts"
    AUTHORED = "authored"
    REFERENCES = "references"
    SIMILAR_TO = "similar_to"
    CONFLICTS_WITH = "conflicts_with"
    OTHER = "other"


@dataclass(frozen=True)
class Relationship:
    """
    Relationship represents a connection between entities.
    
    Relationships carry confidence scores and provenance,
    supporting graph analysis of connected entities.
    """
    id: str
    version: str
    created_at: str
    updated_at: str
    relationship_type: RelationshipType
    source_ref: Reference
    target_ref: Reference
    status: Status
    classification: ClassificationLevel
    confidence: ConfidenceScore
    provenance: Provenance
    evidence_refs: frozenset = field(default_factory=frozenset)
    properties: dict = field(default_factory=dict)
    tags: frozenset = field(default_factory=frozenset)
    metadata: Metadata = field(default_factory=lambda: dict(EMPTY_METADATA))

    @classmethod
    def create(
        cls,
        relationship_type: RelationshipType,
        source_ref: Reference,
        target_ref: Reference,
        confidence: ConfidenceScore,
        provenance: Provenance,
        evidence_refs: list[Reference] | None = None,
        status: Status = Status.ACTIVE,
        classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
        properties: dict | None = None,
        tags: list[str] | None = None,
        metadata: Metadata | None = None,
    ) -> "Relationship":
        """Create a new Relationship."""
        timestamp = utc_now()
        return cls(
            id=new_id(),
            version="1.0.0",
            created_at=timestamp,
            updated_at=timestamp,
            relationship_type=relationship_type,
            source_ref=source_ref,
            target_ref=target_ref,
            status=status,
            classification=ClassificationLevel(classification),
            confidence=confidence,
            provenance=provenance,
            evidence_refs=frozenset(evidence_refs) if evidence_refs else frozenset(),
            properties=properties or {},
            tags=frozenset(tags) if tags else frozenset(),
            metadata=dict(metadata) if metadata else dict(EMPTY_METADATA),
        )


__all__ = [
    "Relationship",
    "RelationshipType",
]
